### PR TITLE
all: Replace "black" with "ruff format".

### PR DESCRIPTION
**Pull request by jimmo**
- Add config for [tool.ruff.format] to pyproject.toml.
- Update pre-commit to run both ruff and ruff-format (and only check C files when running codeformat.py)
- Update a small number of files that change with ruff's rules.
- Update CI.
- Simplify codeformat.py to remove all the Python-specific logic (just run "ruff format" directly).

_This work was funded through GitHub Sponsors._

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,6 +1,3 @@
-[tool.black]
-line-length = 99
-
```
Comment (dpgeorge) [inline pyproject.toml:3]: Maybe it's worth keeping this here, so black can still be used if that's what the user has installed??  Or do we say that black is no longer supported?  I guess that's technically true because we now strictly follow ruff formatting.

**Comment by dpgeorge** [comment]
For the changes which are just reformatting to ruff's specs, can you please split that out into a separate commit?  Either before or after the change to ruff, whichever makes more sense.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -12,6 +12,7 @@ repos:
         verbose: true
         stages: [commit-msg]
   - repo: https://github.com/charliermarsh/ruff-pre-commit
-    rev: v0.1.0
+    rev: v0.1.2
     hooks:
       - id: ruff
+        id: ruff-format
```
Comment (andrewleech) [inline .pre-commit-config.yaml:18]: I'm pretty sure this needs to be: 

```suggestion
    hooks:
      - id: ruff
      - id: ruff-format
```
Without the `-` it's invalid yaml apparently.

File: pyproject.toml (hunk 1 above)
Comment (jimmo) [reply pyproject.toml:3]: I think black is no longer supported.

**Comment by jimmo** [comment]
> For the changes which are just reformatting to ruff's specs, can you please split that out into a separate commit? Either before or after the change to ruff, whichever makes more sense.

Done.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (jimmo) [reply .pre-commit-config.yaml]: Yep, you're right. Fixed.

File: pyproject.toml (hunk 1 above)
Comment (dpgeorge) [reply pyproject.toml:3]: OK

Comment (andrewleech) [reply pyproject.toml:3]: I've made a similar change on a local project and yes, I found it's really one or the other - even the trivially minor differences between the two formatters means you can't really keep both enabled.

**Comment by dpgeorge** [comment]
It looks like the latest ruff is already at 0.1.3.  Is it worth updating the version here?

Also, I'm not sure how the exclude pattern is supposed to work, but when I run `ruff format .` locally it doesn't reformat anything (good!), but if I run `ruff format tests` it does reformat one of the tests.  And if I mess up the formatting of a tests manually (eg in tests/extmod/) then `ruff format .` will not fix it.  It seems that the whole tests dir is excluded.

Note that I'm using 0.1.3 locally, but testing 0.1.2 it has the same behaviour, the whole tests directory seems excluded.

**Comment by jimmo** [comment]
> Also, I'm not sure how the exclude pattern is supposed to work, but when I run `ruff format .` locally it doesn't reformat anything (good!), but if I run `ruff format tests` it does reformat one of the tests. And if I mess up the formatting of a tests manually (eg in tests/extmod/) then `ruff format .` will not fix it. It seems that the whole tests dir is excluded.

You're right -- this config (inadvertently) excludes tests from formatting by default. I don't think this is possible to fix.

- `ruff format` must inherit `ruff check`'s config.
- `ruff check` is the top-level config which can set `exclude`, `extend-exclude`, `include`, `extend-include`.
- `ruff format` can only set `exclude` which appends to to the combined exclude list from `ruff check`.
- We need to exclude tests from `ruff check`, but include them (with specific exceptions) for `ruff format`.
- When you run `ruff format .` it uses the pyproject.toml directly.
- When you run `ruff format tests` it adds `tests` to the include list from pyproject.toml, which takes precedence over the exclude list. (Unless `--force-exclude` is set, see below)

In summary - there doesn't appear to be a way to only exclude a path for `ruff check`. If we could add a `include` to `[tool.ruff.format]` then I think it might work, but that's not currently allowed. I'm not sure I'm up for learning Rust right now. (I think what we really want is for `ruff check` to have its own configuration section).

The best workaround I can think of is to have codeformat.py and ruff.yaml all invoke `ruff format` as `ruff format . tests`, which has the effect of adding tests to the include path only for `ruff format` (in lieu of being able to do this in `[tool.ruff.format]`).

However, `pre-commit` always includes `--force-exclude`, so this won't work there (and I think this would also prevent it working even if `include` was added to `[tool.ruff.format]`). :sob:

**Comment by jimmo** [comment]
For completeness, the other option here is to use `#ruff: noqa` directives to explicitly ignore files/lines so that it's safe to run `ruff format` and `ruff check` directly on the tests directly, but I imagine there are tests that adding the extra comment to will break/change-behaviour. Although maybe there's a small enough number of them that we can add explicit ignores?

**Comment by dpgeorge** [comment]
I guess the simplest option is to treat the tests directory uniformly in terms of linting and formatting.  Linting the tests is probably a good idea anyway, to help find other issues.

Taking a quick look now at how many linting errors there are in `tests/**`, there are quite a few, and it'd take some time to go through each one carefully.

So we could start with excluding most of the subdirs in tests, and include only a few in both linting and formatting.  Then gradually include more and more subdirs.

**Comment by jimmo** [comment]
I found a workaround using per-file-excludes, which means that the tests will still be linted but all lint errors will be ignored.
